### PR TITLE
component/otelcol: set up groundwork for OpenTelemetry Collector components

### PR DESCRIPTION
**Pull request by rfratto**
This commit introduces the component/otelcol package, which will eventually hold a collection of `otelcol.*` components.

Initial prototyping work for Flow OpenTelemetry Collector components was done in #1843, which demonstrated that the full set of code needed to implement OpenTelemetry Components is quite large. I will be splitting up the needed code across a few changes; this is the first.

This initial commit starts setting up the framework for running OpenTelemetry Collector components inside of Flow with a `Scheduler` struct.

Related to #2213.

### CODE REVIEW

File: component/otelcol/host.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	otelcomponent "go.opentelemetry.io/collector/component"
+	otelconfig "go.opentelemetry.io/collector/config"
+)
+
+// host implements otelcomponent.Host for Grafana Agent Flow.
+type host struct {
+	log log.Logger
+
+	extensions map[otelconfig.ComponentID]otelcomponent.Extension
+	exporters  map[otelconfig.DataType]map[otelconfig.ComponentID]otelcomponent.Exporter
```
Comment (mattdurham) [inline component/otelcol/host.go:16]: Thoughts on adding more structure to this? Maps in maps always feel off to me.

Comment (rfratto) [reply component/otelcol/host.go:16]: 🤷 It's the type OpenTelemetry requires for the Host interface. I can name the type but it still needs to be a map of maps.

Comment (mattdurham) [reply component/otelcol/host.go]: Fair enough.

**Review by mattdurham — APPROVED** [review]
LGTM

**Comment by rfratto** [comment]
@tpaschalis I'm going to merge this so I can open another PR, but if you want to review and have feedback, feel free and I'll address tomorrow :)

File: component/otelcol/internal/scheduler/scheduler.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+// Scheduler implements manages a set of OpenTelemetry Collector components.
+// Scheduler is intended to be used from Flow components which need to schedule
+// OpenTelemetry Collector components; it does not implement the full
+// component.Component interface.
+//
+// Each OpenTelemetry Collector component has one instance per supported
+// telemetry signal, hence supporting multiple OpenTelemetry Collector
+// components inside the scheduler. Scheduler should only be used to manage
+// multiple instances of the same OpenTelemetry Collector component.
```
Comment (tpaschalis) [inline component/otelcol/internal/scheduler/scheduler.go:26]: @rfratto I'll read on and maybe things get clearer, but I'm trying to understand what we mean here.

An OpenTelemetry Collector components can support all three of metrics/logs/traces at the same time, right? What do we mean by the following phrase? Should we use one scheduler for multiple OTel Collector component for logs, and then for example, _another_ scheduler for the rest of OTel Collector component for traces+metrics?

> Scheduler should only be used to manage multiple instances of the _same_ OpenTelemetry Collector component.

File: component/otelcol/internal/scheduler/scheduler_test.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		// should cause our running component to terminate.
+		require.NoError(t, started.Wait(5*time.Second), "component did not start")
+		cancel()
+		require.NoError(t, stopped.Wait(5*time.Second), "component did not shutdown")
+	})
+}
+
+func newTriggerComponent() (component otelcomponent.Component, started, stopped *util.WaitTrigger) {
+	started = util.NewWaitTrigger()
+	stopped = util.NewWaitTrigger()
```
Comment (tpaschalis) [inline component/otelcol/internal/scheduler/scheduler_test.go:92]: Just wanted to say this looked nice :)

File: component/otelcol/internal/scheduler/scheduler.go (hunk 2 above)
Comment (rfratto) [reply component/otelcol/internal/scheduler/scheduler.go:26]: Ah, sorry, I'll try to make this clearer over in #2227. 

It probably should say something like: 

> Schedule should only be use to manage the various telemetry signals of the same OpenTelemetry Collector Component.

Would that make more sense? (for example, use the schedule to manage the logs instance, metrics instance, and traces instance for `otelcol.receiver.otlp`)

Comment (tpaschalis) [reply component/otelcol/internal/scheduler/scheduler.go:26]: Is "otlp" here the label, right? So if we have an `otelcol.receiver.prod` and `otelcol.receiver.dev`, we'd have to have one scheduler for each?

Yes then, it makes sense in that context.

Comment (rfratto) [reply component/otelcol/internal/scheduler/scheduler.go:26]: `otelcol.receiver.otlp` is the component name :) But yes, that still applies. I'll make some tweaks in #2227 and ping you there.

**Pull request merged**